File: Solutions/416_Partition_Equal_Subset_Sum/416_Partition_Equal_Subset_Sum.rb
def can_partition(nums)
  sum = nums.inject(0, &:+)
  n = nums.size
  dp = Array.new(n+1) { Array.new(sum+1, false) }
  i = 0
  while i <= n
    dp[i][0] = true
    i+=1
  end
  i = 1
  while i <= n
    j = 1
    while j <= sum/2
      dp[i][j] = dp[i-1][j]
      if nums[i-1] <= j
        dp[i][j] |= dp[i-1][j-nums[i-1]]
      end
      j+=1
    end
    i+=1
  end
  j = sum/2
  return dp[n][j] == true && sum - j == j
end

#Second fastest
def can_partition(nums)
    return false if nums.sum % 2 == 1
    
    @memo = Array.new(nums.size) { Array.new(nums.sum / 2 + 1) }
    return dfs(nums, nums.sum / 2, 0)
end

def dfs(nums, sum, idx)
    return true if sum == 0
    return false if idx >= nums.size || sum < 0
    
    if (val = @memo[idx][sum]) != nil 
        return val
    end
    
    result = dfs(nums, sum - nums[idx], idx + 1) || dfs(nums, sum, idx + 1)
    @memo[idx][sum] = result
end

File: Solutions/416_Partition_Equal_Subset_Sum/416_Partition_Equal_Subset_Sum.go
//https://leetcode.com/problems/partition-equal-subset-sum/discuss/309715/golang
func canPartition(nums []int) bool {
	if len(nums) <= 1 {
		return false
	}
	sum := 0
	for _, num := range nums {
		sum += num
	}

	if sum%2 == 1 {
		return false
	}
	target := sum / 2
	dp := make([]bool, target+1)
	dp[0] = true
	for _, num := range nums {
		for i := target; i >= 1; i-- {
			if i-num >= 0 && dp[i-num] {
				dp[i] = true
				if i == target {
					return true
				}
			}
		}
	}

	return false
}

//Another
func canPartition(nums []int) bool {
	var sum int
	for _, num := range nums {
		sum += num
	}
	if sum%2 == 1 {
		return false
	}
	sum = sum / 2

	dp := make([]bool, sum+1)
	dp[0] = true
	for _, num := range nums {
        if dp[sum]{
            return true
        }
		for j := sum; j > 0; j-- {
			if j >= num && dp[j-num] {
				dp[j] = true
			}
		}
	}
	return dp[sum]
}

//Second fastest
unc canPartition(nums []int) bool {
    var sum int
    for _, n := range nums {
        sum += n
    }
    
    if sum%2 != 0 {
        return false
    }
    
    cache := make(map[int]bool)
    return partition(nums, 0, sum/2, cache)
}

func partition(nums []int, i int, target int, cache map[int]bool) bool {
    if i == len(nums) {
        return false
    }
    if target == 0 {
        return true
    }
    
    k := key(i, target)
    if v, ok := cache[k]; ok {
        return v
    }
 
    diff := target - nums[i]
    if diff >= 0 {
        if partition(nums, i+1, diff, cache) {
            cache[k] = true
            return true
        }
    }
    
    if partition(nums, i+1, target, cache) {
        cache[k] = true
        return true
    }
    
    cache[k] = false
    return false
}

func key(i, target int) int {
    return i*100 + target
}
// 5
// 11
// 5, 11

// 5
// 5, 11
// 5, 5
// 5, 11, 5

// [1,5,11,5]
// [1,5,11,5]
// [1,5,11,5]

//Fastest
func canPartition(nums []int) bool {
    sum := 0
    for _, v := range nums {
        sum += v
    }
    if sum%2 != 0 {
        return false
    }
    sum /= 2
    return dfs(nums, 0, sum)
}

func dfs(nums []int, idx int, target int) bool {
    if target == 0 {
        return true
    }
    if idx == len(nums) || target < 0 {
        return false
    }
    if dfs(nums, idx+1, target-nums[idx]) {
        return true
    }
    k := idx+1
    for k<len(nums) && nums[idx] == nums[k] {
        k++
    }
    return dfs(nums, k, target)
}